### DISCUSSION TITLE
自定义layout组件，在layout组件内如何根据config的routes实现菜单的渲染呢？ props为空

### DISCUSSION THREAD

**Discussion by Cooper1886** [umijs/umi#10175] [Q&A]
项目中有多种layout的需求，那么就需要自定义layout组件，在layout组件内自己实现头部，菜单，底部。现在的问题在于layout组件的props是空的拿不到config的routes，这种情况下如何拿到routes并实现菜单的渲染呢？

**Comment by fz6m** [comment]
通过 `useAppData` 可以拿到路由信息。

如果不需要 layout ，可以配置 `layout: false` ，然后通过 [wrappers](https://umijs.org/docs/guides/routes#wrappers) 实现不同组件对其包裹，在 wrapper 中通过 `<Outlet />` 进行渲染子路由信息，可参考：

https://github.com/umijs/umi/blob/4d283882ca76ae6aa1ac2c25b015ac0c272182ab/examples/boilerplate/.umirc.ts#L27

**Reply by Cooper1886** [reply]
感谢

**Comment by ohroy** [comment]
可以参考这个
https://github.com/umijs/umi/blob/master/packages/plugins/src/layout.ts
通过`useAppData`拿到路由信息，自己遍历下就行了，不过这里的id有点小问题，所以我临时用了name
结合access还可以集成权限判断，参考如下，
layout/user.tsx
```ts
const {clientRoutes, pluginManager} = useAppData()
const newRoutes = clientRoutes.filter((r: any) => r.name === 'userLayout');
const [route] = useAccessMarkedRoutes(mapRoutes(newRoutes));
```

**Reply by Cooper1886** [reply]
感谢
